test/test_opaque_obj_v2.py

```diff
@@ -553,6 +619,157 @@ def test_invalid_schema(self):
                 lib=self.lib,
             )
 
+    def test_invalid_opaque_obj_types(self):
+        for t in [str, bool, int, float, torch.Tensor]:
+            with self.assertRaisesRegex(ValueError, "Unable to register built-in type"):
+                register_opaque_type(t)
+
+        @dataclass
+        class Bad1:
+            x: int
+
+        pytree.register_dataclass(Bad1)
+        with self.assertRaisesRegex(
+            ValueError, "cannot be registered as an opaque object"
+        ):
+            register_opaque_type(Bad1)
```

Comment:
do you handle the case where we do register_opaque_type(Foo) and then Foo goes out of scope correctly? The footgun to avoid is that we shouldn't use id(Foo) in a dict somewhere, and we also probably want a weakref to Foo (instead of a strong reference to it)